File: 921/MinimumAddToMakeParenthesisValid.java
/*
https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/?envType=daily-question&envId=2024-10-09
*/

class Solution {
    public int minAddToMakeValid(String s) {
        int lengthOfS=s.length(), minMovesRequired=0;
        Stack<Character>stack=new Stack<>();
        for(int i=0; i<lengthOfS; ++i){
            char currentChar=s.charAt(i);
            if(currentChar=='('){
                stack.push(currentChar);
            } else{
                if(stack.empty()){
                    minMovesRequired++;
                } else{
                    stack.pop();
                }
            }
        }
        if(!stack.empty()) minMovesRequired+=stack.size();
        return minMovesRequired;
    }
}

File: 9/921/MinimumAddToMakeParenthesisValid.cpp
/*
https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/?envType=daily-question&envId=2024-10-09
*/

class Solution {
public:
    int minAddToMakeValid(string s) {
        int sz=s.size(), res=0;
        stack<char>stk;
        for(auto &ch: s){
            if(ch=='('){
                stk.push('(');
            } else{
                if(stk.empty()){
                    res++;
                } else{
                    stk.pop();
                }
            }
        }
        if(!stk.empty()) res+=stk.size();
        return res;
    }
};